xds/src/main/java/io/grpc/xds/XdsClientImpl.java

```diff
@@ -0,0 +1,725 @@
+   */
+  private void handleRdsResponse(DiscoveryResponse rdsResponse) {
+    logger.log(Level.FINE, "Received an RDS response: {0}", rdsResponse);
+    checkState(adsStream.rdsResourceName != null,
+        "Never requested for RDS resources, management server is doing something wrong");
+    adsStream.rdsRespNonce = rdsResponse.getNonce();
+
+    // Unpack RouteConfiguration messages.
+    List<RouteConfiguration> routeConfigs = new ArrayList<>(rdsResponse.getResourcesCount());
+    try {
+      for (com.google.protobuf.Any res : rdsResponse.getResourcesList()) {
+        routeConfigs.add(res.unpack(RouteConfiguration.class));
+      }
+    } catch (InvalidProtocolBufferException e) {
+      adsStream.sendNackRequest(ADS_TYPE_URL_RDS, ImmutableList.of(adsStream.rdsResourceName),
+          rdsResponse.getNonce(), "Broken RDS response");
+      configWatcher.onError(Status.fromThrowable(e).augmentDescription("Broken RDS response"));
+      return;
+    }
+
+    // Validate and cache information from each RouteConfiguration message.
+    Map<String, String> clusterNames = new HashMap<>();
+    for (RouteConfiguration routeConfig : routeConfigs) {
+      String clusterName = processRouteConfig(routeConfig);
+      if (clusterName == null) {
+        adsStream.sendNackRequest(ADS_TYPE_URL_RDS, ImmutableList.of(adsStream.rdsResourceName),
+            rdsResponse.getNonce(),
+            "Received RouteConfigurations without cluster information for requested hostname");
+        return;
+      }
+      clusterNames.put(routeConfig.getName(), clusterName);
+    }
+    routeConfigNamesToClusterNames.putAll(clusterNames);
+
+    adsStream.sendAckRequest(ADS_TYPE_URL_RDS, ImmutableList.of(adsStream.rdsResourceName),
+        rdsResponse.getVersionInfo(), rdsResponse.getNonce());
+
+    // Notify the ConfigWatcher if this RDS response contains the most recently requested
+    // RDS resource.
+    if (clusterNames.containsKey(adsStream.rdsResourceName)) {
+      ConfigUpdate configUpdate =
+          ConfigUpdate.newBuilder()
+              .setClusterName(clusterNames.get(adsStream.rdsResourceName))
+              .build();
+      configWatcher.onConfigChanged(configUpdate);
+    }
+    // Do not notify an error to the ConfigWatcher. RDS protocol is incremental, not receiving
+    // requested RouteConfiguration in this response does not imply absence.
+  }
+
+  /**
+   * Processes RouteConfiguration message (from an resource information in an LDS or RDS
+   * response), which may contain a VirtualHost with domains matching the "xds:"
+   * URI hostname directly in-line. Returns the clusterName found in the VirtualHost
+   * message with domains matching the "xds:" URI hostname. Returns {@code null} if such a
+   * clusterName cannot be resolved.
+   *
+   * <p>Note we only validate VirtualHosts with domains matching the "xds:" URI hostname as we
+   * never care about others.
```

Comment:
This comment mentions '...domain matching the "xds:" URI hostname...' 3 times which is unnecessary redundancy. Can make the comment concise ...

Reply:
Rephrased. Thanks.